File: packages/core/src/providers/OpenAIProvider.ts
import type { ProviderConfig } from "../models/ProviderConfig.js";
import type { UsageSnapshot } from "../models/UsageSnapshot.js";
import type { IUsageProvider } from "./IUsageProvider.js";

const API_BASE = "https://api.openai.com/v1";
const DEFAULT_LIMIT = 1_000_000; // 1M tokens как ориентир

interface UsageResult {
  input_tokens?: number;
  output_tokens?: number;
}

interface UsageBucket {
  result?: UsageResult | UsageResult[];
}

interface CompletionsResponse {
  data?: UsageBucket[];
}

export class OpenAIProvider implements IUsageProvider {
  validateConfig(config: ProviderConfig): boolean {
    return config.type === "openai" && !!config.apiKey;
  }

  async fetchUsage(config: ProviderConfig): Promise<UsageSnapshot> {
    if (!this.validateConfig(config) || !config.apiKey) {
      return this.errorSnapshot(config.id, "Invalid config");
    }
    const now = Math.floor(Date.now() / 1000);
    const startOfMonth = new Date();
    startOfMonth.setDate(1);
    startOfMonth.setHours(0, 0, 0, 0);
    const startTime = Math.floor(startOfMonth.getTime() / 1000);
    const url = `${API_BASE}/organization/usage/completions?start_time=${startTime}&end_time=${now}&bucket_width=1d`;
    try {
      const res = await fetch(url, {
        headers: { Authorization: `Bearer ${config.apiKey}` },
      });
      if (!res.ok) {
        const text = await res.text();
        return this.errorSnapshot(config.id, `${res.status}: ${text}`);
      }
      const body = (await res.json()) as CompletionsResponse;
      let used = 0;
      for (const bucket of body.data ?? []) {
        const raw = bucket.result;
        const results = Array.isArray(raw) ? raw : raw ? [raw] : [];
        for (const r of results) {
          used += (r?.input_tokens ?? 0) + (r?.output_tokens ?? 0);
        }
      }
      const limit = config.manualLimit ?? DEFAULT_LIMIT;
      const ratio = limit > 0 ? used / limit : 0;
      const status = ratio >= 1 ? "error" : ratio >= 0.8 ? "warning" : "ok";
      return {
        providerId: config.id,
        used,
        limit,
        unit: "tokens",
        timestamp: Date.now(),
        status,
        displayName: config.displayName ?? "OpenAI",
      };
    } catch (err) {
      return this.errorSnapshot(
        config.id,
        err instanceof Error ? err.message : String(err),
      );
    }
  }

  private errorSnapshot(providerId: string, message: string): UsageSnapshot {
    return {
      providerId,
      used: 0,
      limit: 1,
      unit: "tokens",
      timestamp: Date.now(),
      status: "error",
      displayName: "OpenAI",
    };
  }
}

File: apps/android/app/src/main/java/com/ailimits/dashboard/provider/OpenAIProvider.kt
package com.ailimits.dashboard.provider

import com.ailimits.dashboard.data.ProviderConfig
import com.ailimits.dashboard.data.UsageSnapshot
import com.ailimits.dashboard.data.UsageStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL
import javax.net.ssl.HttpsURLConnection

class OpenAIProvider : UsageProvider {
    override fun validateConfig(config: ProviderConfig): Boolean =
        config.type == "openai" && !config.apiKey.isNullOrBlank()

    override suspend fun fetchUsage(config: ProviderConfig): UsageSnapshot =
        withContext(Dispatchers.IO) {
            if (!validateConfig(config) || config.apiKey == null) return@withContext errorSnapshot(config.id)
            val now = System.currentTimeMillis() / 1000
            val startOfMonth = java.util.Calendar.getInstance().apply {
                set(java.util.Calendar.DAY_OF_MONTH, 1)
                set(java.util.Calendar.HOUR_OF_DAY, 0)
                set(java.util.Calendar.MINUTE, 0)
                set(java.util.Calendar.SECOND, 0)
                set(java.util.Calendar.MILLISECOND, 0)
            }.timeInMillis / 1000
            val urlStr =
                "https://api.openai.com/v1/organization/usage/completions?start_time=$startOfMonth&end_time=$now&bucket_width=1d"
            try {
                val url = URL(urlStr)
                (url.openConnection() as HttpsURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("Authorization", "Bearer ${config.apiKey}")
                    connectTimeout = 10000
                    readTimeout = 10000
                }.run {
                    if (responseCode != 200) return@withContext errorSnapshot(config.id)
                    val body = inputStream.bufferedReader().use { it.readText() }
                    val json = JSONObject(body)
                    val data = json.optJSONArray("data") ?: return@withContext errorSnapshot(config.id)
                    var used = 0L
                    for (i in 0 until data.length()) {
                        val bucket = data.optJSONObject(i) ?: continue
                        val result = bucket.opt("result") ?: continue
                        when (result) {
                            is org.json.JSONArray -> {
                                for (j in 0 until result.length()) {
                                    val r = result.optJSONObject(j)
                                    used += (r?.optInt("input_tokens", 0) ?: 0).toLong() +
                                        (r?.optInt("output_tokens", 0) ?: 0).toLong()
                                }
                            }
                            is JSONObject -> {
                                used += (result.optInt("input_tokens", 0)).toLong() +
                                    (result.optInt("output_tokens", 0)).toLong()
                            }
                        }
                    }
                    val limit = config.manualLimit?.toLong() ?: 1_000_000L
                    val ratio = if (limit > 0) used.toDouble() / limit else 0.0
                    UsageSnapshot(
                        providerId = config.id,
                        used = used.toDouble(),
                        limit = limit.toDouble(),
                        unit = "tokens",
                        timestamp = System.currentTimeMillis(),
                        status = usageStatus(ratio),
                        displayName = config.displayName ?: "OpenAI",
                    )
                }
            } catch (_: Exception) {
                errorSnapshot(config.id)
            }
        }

    private fun errorSnapshot(providerId: String) = UsageSnapshot(
        providerId = providerId,
        used = 0.0,
        limit = 1.0,
        unit = "tokens",
        timestamp = System.currentTimeMillis(),
        status = UsageStatus.ERROR,
        displayName = "OpenAI",
    )
}
